beater/v2_handler.go

```diff
@@ -0,0 +1,297 @@
+
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/utility"
+
+	"github.com/santhosh-tekuri/jsonschema"
+
+	"github.com/elastic/apm-server/validation"
+
+	"github.com/elastic/apm-server/decoder"
+	er "github.com/elastic/apm-server/model/error"
+	"github.com/elastic/apm-server/model/metadata"
+	"github.com/elastic/apm-server/model/metric"
+	"github.com/elastic/apm-server/model/span"
+	"github.com/elastic/apm-server/model/transaction"
+)
+
+var (
+	errUnrecognizedObject    = errors.New("did not recognize object type")
+	errInvalidMetadataFormat = errors.New("invalid metadata format")
+)
+
+const batchSize = 10
+
+var models = []struct {
+	key          string
+	schema       *jsonschema.Schema
+	modelDecoder func(interface{}, error) (transform.Transformable, error)
+}{
+	{
+		"transaction",
+		transaction.ModelSchema(),
+		transaction.DecodeEvent,
+	},
+	{
+		"span",
+		span.ModelSchema(),
+		span.DecodeSpan,
+	},
+	{
+		"metric",
+		metric.ModelSchema(),
+		metric.DecodeMetric,
+	},
+	{
+		"error",
+		er.ModelSchema(),
+		er.DecodeEvent,
+	},
+}
+
+type v2Route struct {
+	routeType
+}
+
+func (v v2Route) Handler(beaterConfig *Config, report reporter) http.Handler {
+	reqDecoder := v.configurableDecoder(
+		beaterConfig,
+		func(*http.Request) (map[string]interface{}, error) { return map[string]interface{}{}, nil },
+	)
+
+	v2Handler := v2Handler{
+		requestDecoder: reqDecoder,
+		tconfig:        v.transformConfig(beaterConfig),
+	}
+
+	return v.wrappingHandler(beaterConfig, v2Handler.Handle(beaterConfig, report))
+}
+
+type v2Handler struct {
+	requestDecoder decoder.ReqDecoder
+	tconfig        transform.Config
+}
+
+// handleRawModel validates and decodes a single json object into its struct form
+func (v *v2Handler) handleRawModel(rawModel map[string]interface{}) (transform.Transformable, error) {
+	for _, model := range models {
+		if entry, ok := rawModel[model.key]; ok {
+			err := validation.Validate(entry, model.schema)
+			if err != nil {
+				return nil, err
+			}
+
+			tr, err := model.modelDecoder(entry, err)
+			if err != nil {
+				return tr, err
+			}
+			return tr, nil
+		}
+	}
+	return nil, errUnrecognizedObject
+}
+
+// readBatch will read up to `batchSize` objects from the ndjson stream
+// it returns a slice of eventables, a serverResponse and a bool that indicates if we're at EOF.
+func (v *v2Handler) readBatch(batchSize int, reader *decoder.NDJSONStreamReader, response *StreamResponse) ([]transform.Transformable, bool) {
+	var err error
+	var rawModel map[string]interface{}
+
+	eventables := []transform.Transformable{}
+	for i := 0; i < batchSize && err == nil; i++ {
+		rawModel, err = reader.Read()
+		if err != nil && err != io.EOF {
+			switch e := err.(type) {
+			case decoder.ReadError:
+				response.AddWithMessage(ServerError, 1, e.Error())
+				// return early, we can't recover from a read error
+				return eventables, true
```

Comment:
I was thinking that if there are events that we already read and validated we should `report` them.

Reply:
Fair enough to publish those events.

Out of curiosity - why did you decide to read and publish in batches instead of one by one?

Reply:
good question. With v1, entire payload are put into our queue. There's overhead in pushing to go queues and reading from them, so i wanted to make sure we where not paying that overhead for each even, especially because that wasn't an overhead we had in v1.